# Ebay Car Sale Data - Dataquest Practice Project

This project focuses analysing the data of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.

The original data set can be downloaded from [Kaggle](https://www.kaggle.com/orgesleka/used-cars-database/data). The data used in the analysis is a sample of the orignal data and has been modified by Dataquest for practice purposes.

The data set includes 50,000 data points with the following information:

|column|description|
|:---|:---|
|dateCrawled|When this ad was first crawled|
|name|Name of the car.|
|seller|Whether the seller is private or a dealer.|
|offerType|The type of listing|
|price|The price on the ad to sell the car.|
|abtest|Whether the listing is included in an A/B test.|
|vehicleType|The vehicle Type.|
|yearOfRegistration|The year in which the car was first registered.|
|gearbox|The transmission type.|
|powerPS|The power of the car in PS.|
|model|The car model name.|
|kilometer|How many kilometers the car has driven.|
|monthOfRegistration|The month in which the car was first registered.|
|fuelType|What type of fuel the car uses.|
|brand|The brand of the car.|
|notRepairedDamage|If the car has a damage which is not yet repaired.|
|dateCreated|The date on which the eBay listing was created.|
|nrOfPictures|The number of pictures in the ad.|
|postalCode|The postal code for the location of the vehicle.|
|lastSeenOnline|When the crawler saw this ad last online.|

pandas and NumPy libraries are first imported and the data set is read in as pandas dataframe

In [1]:
import pandas as pd
import numpy as np

autos = pd.read_csv("autos.csv", delimiter = ",", encoding = "Latin-1")

### This section explores the data set on general level and some basic cleaning is executed

The data set is first explored.

The following data columns seem to have missing values:
- vehicletype
- gearbox
- model
- fuelType
- NotRepairedDamage


In [2]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

The columns names are modified to the preferred snakecase format and the following column names are modified for clarity:

- yearOfRegistration -> registration_year
- monthOfRegistration ->  registration_month
- notRepairedDamage ->  unrepaired_damage
- dateCreated ->  ad_created

In [3]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']
autos.head()


date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Next data columns are explored in more detail.

Observations:

- price column has many prices of $0
- registration_year data column has some incorrect values (non-feasible years)
- power_ps column has many entries with 0 horsepower which is incorrect. Also maximum power of 17,700 horsepower seems to be non-feasible.
- registration month has value 0 for some data points
- nr_of_pictures data column seem to have value 0 for all data points

Data that is in incorrect form:

- price data and number of kilometers (odometer) column is not in numeric form
- date columns (dateCrawled, dateCreated & lastSeen) are in text format


In [4]:
autos.describe(include="all")

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-29 23:42:13  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

In [5]:
print(autos["registration_year"].value_counts())
print('\n')
print(autos["power_ps"].value_counts())
print('\n')
print(autos["registration_month"].value_counts())
print('\n')
print(autos["fuel_type"].value_counts())
print('\n')


2000    3354
2005    3015
1999    3000
2004    2737
2003    2727
        ... 
1931       1
1929       1
1001       1
9996       1
1952       1
Name: registration_year, Length: 97, dtype: int64


0        5500
75       3171
60       2195
150      2046
140      1884
         ... 
650         1
490         1
362         1
153         1
16312       1
Name: power_ps, Length: 448, dtype: int64


0     5075
3     5071
6     4368
5     4107
4     4102
7     3949
10    3651
12    3447
9     3389
11    3360
1     3282
8     3191
2     3008
Name: registration_month, dtype: int64


benzin     30107
diesel     14567
lpg          691
cng           75
hybrid        37
andere        22
elektro       19
Name: fuel_type, dtype: int64




### In this section price and odometer kilometer data is cleaned and analysed

Next price data and number of kilometers (odometer) data is changed into numeric form and the column names are updated accordingly.

In [6]:
autos["price"] = autos["price"].str.replace("$","").str.replace(",", "").astype(int)
autos["odometer"] = autos["odometer"].str.replace("km","").str.replace(",", "").astype(int)
autos.rename(columns={"odometer": "odometer_km"}, inplace = True)

The price and odometer_km columns are explored in more detail to identify non-realistic values.

The odometer_km data seems realistic, but it seems include values in only 13 categories, of which the highest is 150 000 km, which has also by far the most data points. That suggest, that many cars quite likely have much more kilometers than 150 000 km, but that data is thus not available in more detail.

The price data has many 0 values and many extremely large values.

In [7]:
print("Price data:")
print(autos["price"].unique().shape)
print('\n')
print(autos["price"].describe())
print('\n')
print(autos["price"].value_counts().sort_index(ascending = True))
print('\n')
print(autos["price"].value_counts().sort_index(ascending = False).head(30))
print('\n')
print("Odometer km data:")
print(autos["odometer_km"].unique().shape)
print('\n')
print(autos["odometer_km"].describe())
print('\n')
print(autos["odometer_km"].value_counts().sort_index(ascending = True))
print('\n')

Price data:
(2357,)


count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


0           1421
1            156
2              3
3              1
5              2
            ... 
10000000       1
11111111       2
12345678       3
27322222       1
99999999       1
Name: price, Length: 2357, dtype: int64


99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
194000      1
190000      1
180000      1
175000      1
169999      1
169000      1
163991      1
163500      1
155000      1
151990      1
Name: price, dtype: int64


Odometer km data:
(13,)


count     50000.000000
mean     125732.700000
std       40042.211706


The large price values reveal some outliers, that do not seem plausible. They include the following data points (name & price). The price for these data points will be replaced with np.nan.

- 39705                        Tausch_gegen_gleichwertiges  99999999
- 42221                                  Leasinguebernahme  27322222
- 39377                        Tausche_volvo_v40_gegen_van  12345678
- 27371                                         Fiat_Punto  12345678
- 47598  Opel_Vectra_B_1_6i_16V_Facelift_Tuning_Showcar...  12345678
- 2897    Escort_MK_1_Hundeknochen_zum_umbauen_auf_RS_2000  11111111
- 24384                            Schlachte_Golf_3_gt_tdi  11111111
- 22947                         Bmw_530d_zum_ausschlachten   1234566
- 43049                                      2_VW_Busse_T3    999999
- 514                     Ford_Focus_Turnier_1.6_16V_Style    999999
- 37585                                Volkswagen_Jetta_GT    999990

In [8]:
print(autos.loc[autos["price"] > 100000, ["name", "price"]].sort_values(by = ["price"], ascending = False))

                                                    name     price
39705                        Tausch_gegen_gleichwertiges  99999999
42221                                  Leasinguebernahme  27322222
39377                        Tausche_volvo_v40_gegen_van  12345678
27371                                         Fiat_Punto  12345678
47598  Opel_Vectra_B_1_6i_16V_Facelift_Tuning_Showcar...  12345678
2897    Escort_MK_1_Hundeknochen_zum_umbauen_auf_RS_2000  11111111
24384                            Schlachte_Golf_3_gt_tdi  11111111
11137  suche_maserati_3200_gt_Zustand_unwichtig_laufe...  10000000
47634                                        Ferrari_FXX   3890000
7814                                         Ferrari_F40   1300000
22947                         Bmw_530d_zum_ausschlachten   1234566
43049                                      2_VW_Busse_T3    999999
514                     Ford_Focus_Turnier_1.6_16V_Style    999999
37585                                Volkswagen_Jetta_GT    99

The outlier price information is replaced with np.nan:

In [9]:
autos.loc[autos["price"] > 10000000, "price"] = np.nan
autos.loc[autos["price"].between(350001, 1299999), "price"] = np.nan

In [10]:
print(autos.loc[autos["price"].isnull(), ["name", "price"]])

                                                    name  price
514                     Ford_Focus_Turnier_1.6_16V_Style    NaN
2897    Escort_MK_1_Hundeknochen_zum_umbauen_auf_RS_2000    NaN
22947                         Bmw_530d_zum_ausschlachten    NaN
24384                            Schlachte_Golf_3_gt_tdi    NaN
27371                                         Fiat_Punto    NaN
37585                                Volkswagen_Jetta_GT    NaN
39377                        Tausche_volvo_v40_gegen_van    NaN
39705                        Tausch_gegen_gleichwertiges    NaN
42221                                  Leasinguebernahme    NaN
43049                                      2_VW_Busse_T3    NaN
47598  Opel_Vectra_B_1_6i_16V_Facelift_Tuning_Showcar...    NaN


In [11]:
print(autos.loc[autos["price"] == 0, ["name", "price", "odometer_km"]].head(20))

                                                  name  price  odometer_km
27              Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE    0.0       150000
71   Suche_Opel_Astra_F__Corsa_oder_Kadett_E_mit_Re...    0.0         5000
80   Nissan_Primera_Hatchback_1_6_16v_73_Kw___99Ps_...    0.0       150000
87                       Bmw_520_e39_zum_ausschlachten    0.0       150000
99                     Peugeot_207_CC___Cabrio_Bj_2011    0.0        60000
118  VW_Sharan_V6_204_PS_Karosse_Rohkarosse_mit_Pap...    0.0       150000
146                                    Ford_Fiesta_rot    0.0        20000
167  Suche_VW_Multivan_Innenausstattung_Set_oder_TE...    0.0         5000
180                                       Zu_verkaufen    0.0       150000
226                      Porsche_911_S_Targa__67er_SWB    0.0         5000
234                                   Fiat_Punto_Sport    0.0       150000
248                             VW_Passat_zu_verkaufen    0.0       150000
259                      

The price of 0 seems also non plausible, so those will also be relaced with np.nan

In [12]:
autos.loc[autos["price"] == 0, "price"] = np.nan

The price data is then checked and some summary statistics is then analysed.

The asking prices for cars range from 1 to 10 000 000, but average asking price is only around 630 and median asking price even lower being 300.

In [13]:
print(autos["price"].value_counts(dropna = False).sort_index(ascending = False).head(10))
print('\n')
print(autos["price"].value_counts(dropna = False).sort_index(ascending = True).head(10))
print('\n')
print(autos["price"].describe())

10000000.0    1
3890000.0     1
1300000.0     1
350000.0      1
345000.0      1
299000.0      1
295000.0      1
265000.0      1
259000.0      1
250000.0      1
Name: price, dtype: int64


1.0     156
2.0       3
3.0       1
5.0       2
8.0       1
9.0       1
10.0      7
11.0      2
12.0      3
13.0      2
Name: price, dtype: int64


count    4.856800e+04
mean     6.201329e+03
std      4.983650e+04
min      1.000000e+00
25%      1.200000e+03
50%      3.000000e+03
75%      7.490000e+03
max      1.000000e+07
Name: price, dtype: float64


For odometer kilometers the data availabie is not that accurate missing more specific information about the large values. Thus the summary statistics are not reliable.

In [14]:
autos["odometer_km"].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

### This section focuses on date and time data

In [15]:
print(autos[["date_crawled", "ad_created", "last_seen", "registration_year", "registration_month"]].head())

          date_crawled           ad_created            last_seen  \
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54   
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08   
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37   
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28   
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50   

   registration_year  registration_month  
0               2004                   3  
1               1997                   6  
2               2009                   7  
3               2007                   6  
4               2003                   7  


For date crawled, ad created, and last seen data only the data about dates are selected for analysis (time data is dropped).

Observations of date crawled data:
- the data is crawled roughly over a time horizon of one month between the 5th of March 2016 and the 7th of April 2016

Observations of ad created data:
- most ads seem to be have been created during the dates when data has been crawled, suggesting that ads are quite short lived possibly due to quick purchases.
- the oldest ad was created almost a year earlier on the 11th of June 2015.

Observations of last seen data:
- only around 13% of ads were last seen on the last date when data was crawled reinforcing the idea, that ads are on average quite short lived.

In [16]:
date_crawled_dates_only = autos["date_crawled"].str[:10]
print(date_crawled_dates_only.value_counts(normalize = True, dropna = False).sort_index(ascending = True))

2016-03-05    0.02538
2016-03-06    0.01394
2016-03-07    0.03596
2016-03-08    0.03330
2016-03-09    0.03322
2016-03-10    0.03212
2016-03-11    0.03248
2016-03-12    0.03678
2016-03-13    0.01556
2016-03-14    0.03662
2016-03-15    0.03398
2016-03-16    0.02950
2016-03-17    0.03152
2016-03-18    0.01306
2016-03-19    0.03490
2016-03-20    0.03782
2016-03-21    0.03752
2016-03-22    0.03294
2016-03-23    0.03238
2016-03-24    0.02910
2016-03-25    0.03174
2016-03-26    0.03248
2016-03-27    0.03104
2016-03-28    0.03484
2016-03-29    0.03418
2016-03-30    0.03362
2016-03-31    0.03192
2016-04-01    0.03380
2016-04-02    0.03540
2016-04-03    0.03868
2016-04-04    0.03652
2016-04-05    0.01310
2016-04-06    0.00318
2016-04-07    0.00142
Name: date_crawled, dtype: float64


In [17]:
ad_created_dates_only = autos["ad_created"].str[:10]
print(ad_created_dates_only.value_counts(normalize = True, dropna = False).sort_index(ascending = True).head(40))
print(ad_created_dates_only.value_counts(normalize = True, dropna = False).sort_index(ascending = True).tail(40))


2015-06-11    0.00002
2015-08-10    0.00002
2015-09-09    0.00002
2015-11-10    0.00002
2015-12-05    0.00002
2015-12-30    0.00002
2016-01-03    0.00002
2016-01-07    0.00002
2016-01-10    0.00004
2016-01-13    0.00002
2016-01-14    0.00002
2016-01-16    0.00002
2016-01-22    0.00002
2016-01-27    0.00006
2016-01-29    0.00002
2016-02-01    0.00002
2016-02-02    0.00004
2016-02-05    0.00004
2016-02-07    0.00002
2016-02-08    0.00002
2016-02-09    0.00004
2016-02-11    0.00002
2016-02-12    0.00006
2016-02-14    0.00004
2016-02-16    0.00002
2016-02-17    0.00002
2016-02-18    0.00004
2016-02-19    0.00006
2016-02-20    0.00004
2016-02-21    0.00006
2016-02-22    0.00002
2016-02-23    0.00008
2016-02-24    0.00004
2016-02-25    0.00006
2016-02-26    0.00004
2016-02-27    0.00012
2016-02-28    0.00020
2016-02-29    0.00016
2016-03-01    0.00010
2016-03-02    0.00010
Name: ad_created, dtype: float64
2016-02-28    0.00020
2016-02-29    0.00016
2016-03-01    0.00010
2016-03-02    0.00010

In [18]:
last_seen_dates_only = autos["last_seen"].str[:10]
print(last_seen_dates_only.value_counts(normalize = True, dropna = False).sort_index(ascending = True))

2016-03-05    0.00108
2016-03-06    0.00442
2016-03-07    0.00536
2016-03-08    0.00760
2016-03-09    0.00986
2016-03-10    0.01076
2016-03-11    0.01252
2016-03-12    0.02382
2016-03-13    0.00898
2016-03-14    0.01280
2016-03-15    0.01588
2016-03-16    0.01644
2016-03-17    0.02792
2016-03-18    0.00742
2016-03-19    0.01574
2016-03-20    0.02070
2016-03-21    0.02074
2016-03-22    0.02158
2016-03-23    0.01858
2016-03-24    0.01956
2016-03-25    0.01920
2016-03-26    0.01696
2016-03-27    0.01602
2016-03-28    0.02086
2016-03-29    0.02234
2016-03-30    0.02484
2016-03-31    0.02384
2016-04-01    0.02310
2016-04-02    0.02490
2016-04-03    0.02536
2016-04-04    0.02462
2016-04-05    0.12428
2016-04-06    0.22100
2016-04-07    0.13092
Name: last_seen, dtype: float64


Next the regstration year data is analysed. The describtion of the data reveals, that the data has non plausible values (for example min and max values), and should be cleaned.

In [19]:
autos["registration_year"].describe()

count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The data seet seem to have some clearly erraneous registration year values, but also a large amount of registration year values of year 2017 and 2018, even though the data set was crawled in 2016. Quick exploration of the ads with registration year 2017 or 2018 reveals, that the cars do not seem to be new cars. Also cars with registration year less than or equal to year 1910 seem to have incorrect registration year. Based on the findings all registration years with values lower than 1911 or higher than 2016 will be replaced with the value np.nan.

In [20]:
autos.loc[autos["registration_year"] < 1900, "registration_year"].value_counts()
autos.loc[autos["registration_year"] > 2016, "registration_year"].value_counts()

2017    1453
2018     492
9999       4
5000       4
2019       3
9000       2
5911       1
9996       1
2800       1
4800       1
8888       1
4500       1
6200       1
4100       1
Name: registration_year, dtype: int64

In [21]:
print(autos.loc[autos["registration_year"] > 2016, ["name", "registration_year", "odometer_km"]].head(20))

                                                  name  registration_year  \
10                        VW_Golf_Tuning_in_siber/grau               2017   
55                      Mercedes_E320_AMG_zu_Tauschen!               2017   
65                       Ford_Fiesta_zum_ausschlachten               2017   
68   Mini_cooper_s_clubman_/vollausstattung_/_Navi/...               2017   
84                                      Renault_twingo               2018   
113                              Golf_4_Anfaenger_auto               2017   
164  Opel_Meriva__nur_76000_Km__unfallfrei__scheckh...               2018   
197                              VW_Polo_9N_an_Bastler               2017   
253             Ford_mondeo_Gas_anlage_mit_TÜV_04.2017               2017   
348  VW_Beetle_1.8Turbo_mit_Vollausstattung_und_seh...               2017   
390   Fiat_Bertone_X_1_9__X_1/9__X19__X_19__X1_9__X_19               2018   
438                VW_Golf_5_V_1.6_2004_Klima_Tempomat               2017   

In [22]:
print(autos.loc[autos["registration_year"] < 1900, ["name", "registration_year", "odometer_km"]])

                                                   name  registration_year  \
10556                                        UNFAL_Auto               1800   
22316             VW_Kaefer.__Zwei_zum_Preis_von_einem.               1000   
24511                       Trabant__wartburg__Ostalgie               1111   
32585                                        UNFAL_Auto               1800   
35238  Suche_Skoda_Fabia____Skoda_Fabia_Combi_mit_Klima               1500   
49283                                        Citroen_HY               1001   

       odometer_km  
10556         5000  
22316         5000  
24511         5000  
32585         5000  
35238         5000  
49283         5000  


In [23]:
print(autos.loc[autos["registration_year"].between(1900, 1930), ["name", "registration_year", "odometer_km"]])

                                              name  registration_year  \
3679                                    Suche_Auto               1910   
15898  Tausch_alles_aus_meinen_Anzeigen_gegen_Auto               1910   
21416                      Essex_super_six__Ford_A               1927   
22101                                   BMW_Andere               1929   
22659                                 Opel_Corsa_B               1910   
28693                               Renault_Twingo               1910   
30781             Opel_Calibra_V6_DTM_Bausatz_1:24               1910   
33295       Ich_verkaufe/suche_fuer_sie_ihr_Auto!!               1910   
42181    SAMSUNG_55“_3D_Tv_und_Soundbar_gegen_Auto               1910   
45157                                   Motorhaube               1910   
46213                                Bellier_Vario               1910   

       odometer_km  
3679          5000  
15898         5000  
21416         5000  
22101         5000  
22659       150000

Exploring the data without those cars with erreneous registration year, most common registration year of cars are between the years 1997 and 2009.

In [24]:
autos.loc[autos["registration_year"] < 1911, "registration_year"] = np.nan
autos.loc[autos["registration_year"] > 2016, "registration_year"] = np.nan

print(autos["registration_year"].value_counts(dropna = False).sort_index(ascending = True).head(10))
print(autos["registration_year"].value_counts(dropna = False).sort_index(ascending = True).tail(10))
print(autos["registration_year"].describe())
print(autos["registration_year"].value_counts(normalize = True).sort_index(ascending = True).head(40))
print(autos["registration_year"].value_counts(normalize = True).sort_index(ascending = True).tail(40))

1927.0    1
1929.0    1
1931.0    1
1934.0    2
1937.0    4
1938.0    1
1939.0    1
1941.0    2
1943.0    1
1948.0    1
Name: registration_year, dtype: int64
2008.0    2231
2009.0    2098
2010.0    1597
2011.0    1634
2012.0    1323
2013.0     806
2014.0     666
2015.0     399
2016.0    1316
NaN       1981
Name: registration_year, dtype: int64
count    48019.000000
mean      2002.820904
std          7.200279
min       1927.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64
1927.0    0.000021
1929.0    0.000021
1931.0    0.000021
1934.0    0.000042
1937.0    0.000083
1938.0    0.000021
1939.0    0.000021
1941.0    0.000042
1943.0    0.000021
1948.0    0.000021
1950.0    0.000062
1951.0    0.000042
1952.0    0.000021
1953.0    0.000021
1954.0    0.000042
1955.0    0.000042
1956.0    0.000104
1957.0    0.000042
1958.0    0.000083
1959.0    0.000146
1960.0    0.000708
1961.0    0.000125
1962.0    0.000083
1

### In this section the prices of different car brands are analysed

The analysis will focus on top 20 most common car brands (sonstige_autos will be removed)

In [27]:
print(autos["brand"].value_counts().sort_values(ascending = False).head(21))

volkswagen        10687
opel               5461
bmw                5429
mercedes_benz      4734
audi               4283
ford               3479
renault            2404
peugeot            1456
fiat               1308
seat                941
skoda               786
mazda               757
nissan              754
smart               701
citroen             701
toyota              617
sonstige_autos      546
hyundai             488
volvo               457
mini                424
mitsubishi          406
Name: brand, dtype: int64


A list that has the the brands of top 20 most common cars (not including "sonstige_autos") is first created.

In [28]:
top_21_car_brands_value_counts = autos["brand"].value_counts().sort_values(ascending = False).head(21)
print(top_21_car_brands_value_counts)
top_20_car_brands_value_counts = top_21_car_brands_value_counts.drop(labels = ["sonstige_autos"])
top_20_car_brands = top_20_car_brands_value_counts.index
print(top_20_car_brands)
print(len(top_20_car_brands))

volkswagen        10687
opel               5461
bmw                5429
mercedes_benz      4734
audi               4283
ford               3479
renault            2404
peugeot            1456
fiat               1308
seat                941
skoda               786
mazda               757
nissan              754
smart               701
citroen             701
toyota              617
sonstige_autos      546
hyundai             488
volvo               457
mini                424
mitsubishi          406
Name: brand, dtype: int64
Index(['volkswagen', 'opel', 'bmw', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'mazda', 'nissan', 'smart',
       'citroen', 'toyota', 'hyundai', 'volvo', 'mini', 'mitsubishi'],
      dtype='object')
20


Next the list of top 20 brands will be used to calculate the average prices of those brands.

In [29]:
selection = autos[autos["brand"] == "volkswagen"]
print(len(selection["price"]))
print(selection["price"].sum())
print(selection["price"].count())

10687
55116497.0
10336


Of the 20 most common car brand, on average the most expesive are (top 5):
- Mini (10 500)
- Audi (9200)
- Mercedes Benz (8500)
- BMW (8300)
- Skoda (6400)

And on average the cheapest are (bottom 5):
- Renault (2400)
- Fiat (2800)
- Opel (2900)
- Peugeot (3100)
- Misubishi (3400)
 

In [34]:
average_prices_top_20_brands = {}

for brand in top_20_car_brands:
    selected_rows = autos.loc[autos["brand"] == brand]
    mean_price = selected_rows["price"].sum() / selected_rows["price"].count()
    average_prices_top_20_brands[brand] = mean_price
    
sorted_prices = sorted(average_prices_top_20_brands.items(), key=lambda x: x[1], reverse = True) 
print(sorted_prices)
    
    
    

[('mini', 10541.566985645934), ('audi', 9212.9306621881), ('mercedes_benz', 8536.027085124677), ('bmw', 8261.382442169132), ('skoda', 6353.544871794872), ('hyundai', 5371.792960662526), ('volkswagen', 5332.4784249226), ('toyota', 5148.0032733224225), ('volvo', 4866.993166287016), ('nissan', 4669.3859649122805), ('seat', 4315.744565217391), ('mazda', 4059.059539918809), ('citroen', 3761.8950437317785), ('ford', 3728.4121821407452), ('smart', 3518.102305475504), ('mitsubishi', 3372.190476190476), ('peugeot', 3065.611888111888), ('opel', 2944.6075421641085), ('fiat', 2793.8700475435817), ('renault', 2431.195698924731)]


In addition to prices, the avege mileage of the top 20 car brands is calculated.

In [36]:
average_mileage_top_20_brands = {}

for brand in top_20_car_brands:
    selected_rows = autos.loc[autos["brand"] == brand]
    mean_mileage = selected_rows["odometer_km"].sum() / selected_rows["odometer_km"].count()
    average_mileage_top_20_brands[brand] = mean_mileage
    
sorted_mileage = sorted(average_mileage_top_20_brands.items(), key=lambda x: x[1]) 
print(sorted_mileage)

[('mini', 89375.0), ('smart', 100756.06276747503), ('hyundai', 106782.7868852459), ('skoda', 110947.83715012722), ('toyota', 115988.65478119935), ('fiat', 117037.4617737003), ('nissan', 118978.7798408488), ('citroen', 119764.62196861627), ('seat', 122061.63655685441), ('ford', 124131.93446392642), ('mazda', 125132.10039630119), ('mitsubishi', 126293.10344827586), ('peugeot', 127352.33516483517), ('renault', 128223.79367720465), ('volkswagen', 128955.27276129878), ('opel', 129298.66324848929), ('audi', 129643.9411627364), ('mercedes_benz', 130886.14279678918), ('bmw', 132521.64302818198), ('volvo', 138632.3851203501)]


Next the average prices and mileage are stored as a single dataframe.

Based on the average values, the most expensive brand Mini has also on average lowest mileage of the top 20 brands. Also Skoda is on average quite expensive (5th most expensive) and has on average low mileage (4th lowest).

Audi, BMW and Mercedes Benz on the other hand are among the top 5 most expensive brands but at the same time they have the most mileage after Volvo.

(when interpreting the results, it has to be mentioned again, that the mileage data is not very accurate. Especially the top category 150 000 km limits the accuracy of the data)

In [56]:
prices_series = pd.Series(average_prices_top_20_brands)
prices_dataframe = pd.DataFrame(prices_series, columns = ["mean_prices"])

mileage_series = pd.Series(average_mileage_top_20_brands)
prices_dataframe["mean_mileage"] = mileage_series
average_prices_mileage = prices_dataframe
print("Sorted by brand name")
print(average_prices_mileage.sort_index(axis = 0).round(decimals = -2).astype(int))
print('\n')
print("Sorted by average price (most expensive first)")
print(average_prices_mileage.sort_values(by="mean_prices", ascending = False).round(decimals = -2).astype(int))
print('\n')
print("Sorted by average mileage (smallest mileage first)")
print(average_prices_mileage.sort_values(by="mean_mileage", ascending = True).round(decimals = -2).astype(int))

Sorted by brand name
               mean_prices  mean_mileage
audi                  9200        129600
bmw                   8300        132500
citroen               3800        119800
fiat                  2800        117000
ford                  3700        124100
hyundai               5400        106800
mazda                 4100        125100
mercedes_benz         8500        130900
mini                 10500         89400
mitsubishi            3400        126300
nissan                4700        119000
opel                  2900        129300
peugeot               3100        127400
renault               2400        128200
seat                  4300        122100
skoda                 6400        110900
smart                 3500        100800
toyota                5100        116000
volkswagen            5300        129000
volvo                 4900        138600


Sorted by average price (most expensive first)
               mean_prices  mean_mileage
mini                 10500  